Add conversation API to intents section (#1583)
home-assistant/developers.home-assistant: docs/intent_conversation_api.md

---
title: "Conversation API"
sidebar_label: "Conversation API"
---

Intents can be recognized from text and fired using the [conversation integration](https://www.home-assistant.io/integrations/conversation/).

An HTTP API endpoint is available at `/api/conversation/process`, which receives an [input sentence](#input-sentence) and produces an [intent response](#intent-response).

## Input Sentence

A sentence may be POST-ed to `/api/conversation/process` like:

```json
{
  "text": "turn on the lights in the living room",
  "language": "en"
}
```

The following input fields are available:

| Name              | Type   | Description                                                                     |
|-------------------|--------|---------------------------------------------------------------------------------|
| `text`            | string | Input sentence.                                                                 |
| `language`        | string | Optional. Language of the input sentence (defaults to configured language).     |

> | `conversation_id` | string | Optional. Unique id to track conversation across multiple inputs and responses. |

Reviewer: Should we mention that we will return conversation_id to users, and they shouldn't create their own ?

Author: We should, and I need to double check that I'm actually doing that...

## Intent Response

The response from `/api/conversation/process` contains information about the effect of the fired intent, for example:

```json
{
  "response_type": "action_done",
  "language": "en",
  "data": {

>     "target": {

Reviewer: This is missing `id`

      "type": "area",
      "name": "Living Room"
    }
  },
  "speech": {
    "plain": {

>       "speech": "Turn Living Room lights on"

Reviewer: This was the input, it should be the response

| `type` | string | Target type. One of `area`, `device`, or `entity`. |
| `name` | string | Name of the affected area/device/entity.           |
| `id`   | string | Optional. Id of the target.                        |

### Query Answer

The response is an answer to a question, such as "what is the temperature?". See the [speech](#speech) property for the answer text.

```json
{
  "response_type": "query_answer",
  "language": "en",
  "speech": {
    "plain": {
      "speech": "It is 65 degrees"
    }
  }
}
```

### Error

An error occurred either during intent recognition or handling. See `data.code` for the specific type of error, and the [speech](#speech) property for the error message.

```json
{
  "response_type": "error",
  "language": "en",
  "data": {
    "code": "no_intent_match"
  },
  "speech": {
    "plain": {
      "speech": "Sorry, I didn't understand that"
    }
  }
}
```

`data.code` can be one of:

* `no_intent_match` - The input text did not match any intents.
* `no_valid_targets` - The targeted area, device, or entity does not exist.
* `failed_to_handle` - An unexpected error occurred while handling the intent.

> 

Reviewer: Add code `unknown`

| `language`        | string     | Optional. The language of the intent and response.                                                  |
| `speech`          | dictionary | Optional. Response text to speak to the user (see [speech](#speech)).                               |
| `conversation_id` | string     | Optional. Unique id to [track conversation](#conversation-id) across multiple inputs and responses. |

## Response Types

### Action Done

The intent produced an action in Home Assistant, such as turning on a light. The `data` property of the response contains a `targets` list, where each target looks like:

| Name   | Type   | Description                                                                            |
|--------|--------|----------------------------------------------------------------------------------------|
| `type` | string | Target type. One of `area`, `domain`, `device_class`, `device`, `entity`, or `custom`. |
| `name` | string | Name of the affected target.                                                           |
| `id`   | string | Optional. Id of the target.                                                            |

> Targets must be ordered from general to specific:

Reviewer: ```suggestion
An intent can have multiple targets which are applied on top of each other. The targets must be ordered from general to specific:
```

An HTTP API endpoint is available at `/api/conversation/process`, which receives an [input sentence](#input-sentence) and produces an [intent response](#intent-response).
A "conversation" is tracked across multiple inputs and responses using a [conversation id](#conversation-id) generated by Home Assistant.

## Input Sentence

A sentence may be POST-ed to `/api/conversation/process` like:

```json
{
  "text": "turn on the lights in the living room",
  "language": "en"
}
```

The following input fields are available:

| Name              | Type   | Description                                                                                 |
|-------------------|--------|---------------------------------------------------------------------------------------------|
| `text`            | string | Input sentence.                                                                             |
| `language`        | string | Optional. Language of the input sentence (defaults to configured language).                 |
| `conversation_id` | string | Optional. Unique id to [track conversation](#conversation-id). Generated by Home Assistant. |

> ## Intent Response

Reviewer: ```suggestion
## Conversation Response
```

```json
{
  "response": {
    "response_type": "action_done",
    "language": "en",
    "data": {
      "targets": [
        {
          "type": "area",
          "name": "Living Room"
          "id": "living_room"
        },
        {
          "type": "domain",
          "name": "light"
          "id": "light"
        }
      ]
    },
    "speech": {
      "plain": {
        "speech": "Turned Living Room lights on"
      }
    }
  },
  "conversation_id": "<generated-id-from-ha>",
}
```

The following properties are available in the `"response"` object:

| Name              | Type       | Description                                                                                         |
|-------------------|------------|-----------------------------------------------------------------------------------------------------|
| `response_type`   | string     | One of `action_done`, `query_answer`, or `error` (see [response types](#response-types)).           |
| `data`            | dictionary | Relevant data for each [response type](#response_types).                                            |

> | `language`        | string     | Optional. The language of the intent and response.                                                  |

Reviewer: This is not optional.

Author: I need to enforce this in the IntentResponse.

| `speech`          | dictionary | Optional. Response text to speak to the user (see [speech](#speech)).                               |

> The [conversation id](#conversation-id) is returned alongside the intent response.

Reviewer: ```suggestion
The [conversation id](#conversation-id) is returned alongside the intent response. Some conversation agents support tracking context across input so it's important to always send this ID to the API with follow-up requests.
```

      }
    }
  },
  "conversation_id": "<generated-id-from-ha>",
}
```

### Error

An error occurred either during intent recognition or handling. See `data.code` for the specific type of error, and the [speech](#speech) property for the error message.

```json
{
  "response": {
    "response_type": "error",
    "language": "en",
    "data": {
      "code": "no_intent_match"
    },
    "speech": {
      "plain": {
        "speech": "Sorry, I didn't understand that"
      }
    }
  }
}
```

`data.code` is a string that can be one of:

* `no_intent_match` - The input text did not match any intents.
* `no_valid_targets` - The targeted area, device, or entity does not exist.
* `failed_to_handle` - An unexpected error occurred while handling the intent.
* `unknown` - An error occurred outside the scope of intent processing.

## Speech

The spoken response to the user is provided in the `speech` property of the response. It can either be plain text (the default), or [SSML](https://www.w3.org/TR/speech-synthesis11/).

For plain text speech, the response will look like:

```json
{
  "response": {
    "response_type": "...",
    "speech": {
      "plain": {
        "speech": "...",

>         "extra_data": null

Reviewer: if we don't have use for `extra_data` yet, should we just kill it ?

Author: Yeah, any scenario I can think of to use it should probably just be SSML anyways.

      }
    }
  },
  "conversation_id": "<generated-id-from-ha>",
}
```

If the speech is [SSML](https://www.w3.org/TR/speech-synthesis11/), it will instead be:

```json
{
  "response": {
    "response_type": "...",
    "speech": {
      "ssml": {
        "speech": "...",
        "extra_data": null
      }
    }
  },
  "conversation_id": "<generated-id-from-ha>",
}
```

## Conversation Id

> Conversations are tracked by a unique id generated from within Home Assistant. To continue a conversation, retrieve the `conversation_id` from the HTTP API response (alongside the [intent response](#intent-response)) and add it to the next [input sentence](#input-sentence):

Reviewer: ```suggestion
Conversations can be tracked by a unique id generated from within Home Assistant if supported by the answering conversation agent. To continue a conversation, retrieve the `conversation_id` from the HTTP API response (alongside the [intent response](#intent-response)) and add it to the next [input sentence](#input-sentence):
```

| `conversation_id` | string | Optional. Unique id to [track conversation](#conversation-id). Generated by Home Assistant. |

## Conversation Response

The JSON response from `/api/conversation/process` contains information about the effect of the fired intent, for example:

```json
{
  "response": {
    "response_type": "action_done",
    "language": "en",
    "data": {
      "targets": [
        {
          "type": "area",
          "name": "Living Room"
          "id": "living_room"
        },
        {
          "type": "domain",
          "name": "light"
          "id": "light"
        }
      ]
    },
    "speech": {
      "plain": {
        "speech": "Turned Living Room lights on"
      }
    }
  },
  "conversation_id": "<generated-id-from-ha>",
}
```

The following properties are available in the `"response"` object:

| Name            | Type       | Description                                                                                                      |
|-----------------|------------|------------------------------------------------------------------------------------------------------------------|

> | `response_type` | string     | One of `action_done`, `partial_action_done`, `query_answer`, or `error` (see [response types](#response-types)). |

Reviewer: We're removing `partial_action_done`

Reviewer: ```suggestion
| `response_type` | string     | One of `action_done`, `query_answer`, or `error` (see [response types](#response-types)). |
```

| `id`       | string  | Optional. Id of the target.                                                            |

An intent can have multiple targets which are applied on top of each other. The targets must be ordered from general to specific:

* `area`
  * A [registered area](https://developers.home-assistant.io/docs/area_registry_index/)
* `domain`
  * Home Assistant integration domain, such as "light"
* `device_class`
  * Device class for a domain, such as "garage_door" for the "cover" domain
* `device`
  * A [registered device](https://developers.home-assistant.io/docs/device_registry_index)
* `entity`
  * A [Home Assistant entity](https://developers.home-assistant.io/docs/architecture/devices-and-services)

Most intents end up with 0, 1 or 2 targets. 3 targets currenly only happens when device classes are involved. Examples of target combinations:

 * "Turn off all lights"
     * 1 target: `domain:light`
 * "Turn on the kitchen lights"
     * 2 targets: `area:kitchen`, `domain:light`
 * "Open the kitchen blinds"
     * 3 targets: `area:kitchen`, `domain:cover`, `device_class:blind`

> ### Partial Action Done

Reviewer: Merge into `Action Done`
